onnxruntime/core/providers/cuda/math/clip.cc

```diff
@@ -92,25 +86,27 @@ struct LowMax<MLFloat16> {
 template <typename T>
 struct Clip::ComputeImpl {
   void operator()(cudaStream_t stream, const Tensor* X, const Tensor* min, const Tensor* max, Tensor* Y) const {
-    auto min_val = clip_internal::LowMax<T>::low();
-    auto max_val = clip_internal::LowMax<T>::max();
+    auto min_default = clip_internal::LowMax<T>::low();
+    auto max_default = clip_internal::LowMax<T>::max();
 
+    const T* min_data = nullptr;
+    const T* max_data = nullptr;
     // 1-2 Input on CPU
     if (min) {
-      ORT_ENFORCE(min->Shape().NumDimensions() == 0, "min should be a scalar.");
-      min_val = *(min->template Data<T>());
+      ORT_ENFORCE(min->Shape().Size() == 1, "min should be a scalar.");
```

Comment:
Is it necessary to relax the input shape check here (and below)? As far as I can tell, from the [spec](https://github.com/onnx/onnx/blob/master/docs/Changelog.md#Clip-13) it is supposed to have `NumDimensions() == 0`:

> Minimum value, under which element is replaced by min. It must be a scalar(tensor of empty shape).

If we're seeing something else at this point, could it indicate an exporter bug?

Reply:
The input is produced by a where, whose output is a tensor with 1 element. I looked it up and there is at least one other op: [Range](https://github.com/onnx/onnx/blob/master/docs/Operators.md#range) that has inputs that must be a scalar, but the kernel does a [IsScalar()](https://github.com/microsoft/onnxruntime/blob/b370ddbf5ee823d0c1e5e095fd03bf6a290e7b06/onnxruntime/core/providers/cuda/generator/range.cc#L38) check which [relaxes the condition to to be either 0D or 1D with 1 element](https://github.com/microsoft/onnxruntime/blob/b370ddbf5ee823d0c1e5e095fd03bf6a290e7b06/include/onnxruntime/core/framework/tensor_shape.h#L127). I can change the check to match Range's IsScalar() check.